src/nvim/mark_extended.h

```diff
@@ -0,0 +1,267 @@
+      extmark = &kb_itr_key(&mitr);\
+      if (extmark->line->lnum == l_lnum \
+          && extmark->col < l_col) { \
+          break;\
+      }\
+      code;\
+    }\
+  })
+
+
+#define FOR_ALL_EXTMARKS_IN_LINE(items, l_col, u_col, code)\
+  kbitr_t(markitems) mitr;\
+  ExtendedMark mt;\
+  mt.ns_id = 0;\
+  mt.mark_id = 0;\
+  mt.line = NULL;\
+  mt.col = l_col;\
+  colnr_T extline_u_col = u_col;\
+  if (!kb_itr_get(markitems, &items, mt, &mitr)) { \
+    kb_itr_next(markitems, &items, &mitr);\
+  } \
+  ExtendedMark *extmark;\
+  for (; kb_itr_valid(&mitr); kb_itr_next(markitems, &items, &mitr)) { \
+    extmark = &kb_itr_key(&mitr);\
+    if (extmark->col > extline_u_col) { \
+      break;\
+    }\
+    code;\
+  }
+
+
+typedef struct ExtmarkNs {  // For namespacing extmarks
+  PMap(uint64_t) *map;      // For fast lookup
+  uint64_t free_id;         // For automatically assigning id's
+} ExtmarkNs;
+
+
+typedef kvec_t(ExtendedMark *) ExtmarkArray;
+typedef kvec_t(ExtMarkLine *) ExtlineArray;
+
+
+// Undo/redo extmarks
+
+typedef enum {
+  kExtmarkNOOP,        // Extmarks shouldn't be moved
+  kExtmarkUndo,        // Operation should be reversable/undoable
+  kExtmarkNoUndo,      // Operation should not be reversable
+  kExtmarkUndoNoRedo,  // Operation should be undoable, but not redoable
+} ExtmarkOp;
+
+
+typedef struct {
+  linenr_T line1;
+  linenr_T line2;
+  long amount;
+  long amount_after;
+} Adjust;
+
+typedef struct {
+  linenr_T lnum;
+  colnr_T mincol;
+  long col_amount;
+  long lnum_amount;
+} ColAdjust;
+
+typedef struct {
+    linenr_T lnum;
+    colnr_T mincol;
+    colnr_T endcol;
+    int eol;
+} ColAdjustDelete;
+
+typedef struct {
+  linenr_T line1;
+  linenr_T line2;
+  linenr_T last_line;
+  linenr_T dest;
+  linenr_T num_lines;
+  linenr_T extra;
+} AdjustMove;
+
+typedef struct {
+  uint64_t ns_id;
+  uint64_t mark_id;
+  linenr_T lnum;
+  colnr_T col;
+} ExtmarkSet;
+
+typedef struct {
+  uint64_t ns_id;
+  uint64_t mark_id;
+  linenr_T old_lnum;
+  colnr_T old_col;
+  linenr_T lnum;
+  colnr_T col;
+} ExtmarkUpdate;
+
+typedef struct {
+  uint64_t ns_id;
+  uint64_t mark_id;
+  linenr_T lnum;
+  colnr_T col;
+} ExtmarkCopy;
+
+typedef struct {
+  linenr_T l_lnum;
+  colnr_T l_col;
+  linenr_T u_lnum;
+  colnr_T u_col;
+  linenr_T p_lnum;
+  colnr_T p_col;
+} ExtmarkCopyPlace;
+
+typedef struct {
+  uint64_t ns_id;
+  linenr_T l_lnum;
+  linenr_T u_lnum;
+} ExtmarkClear;
+
+
+typedef enum {
+  kLineAdjust,
+  kColAdjust,
+  kColAdjustDelete,
+  kAdjustMove,
+  kExtmarkSet,
+  kExtmarkDel,
+  kExtmarkUpdate,
+  kExtmarkCopy,
+  kExtmarkCopyPlace,
+  kExtmarkClear,
+} UndoObjectType;
+
+struct undo_object {
+  UndoObjectType type;
+  union {
+    Adjust adjust;
+    ColAdjust col_adjust;
+    ColAdjustDelete col_adjust_delete;
+    AdjustMove move;
+    ExtmarkSet set;
+    ExtmarkUpdate update;
+    ExtmarkCopy copy;
+    ExtmarkCopyPlace copy_place;
+    ExtmarkClear clear;
+  } data;
```

Comment:
this seems extreme, I wonder if we could have just 1 `Adjust` type with various fields.

Reply:
I agree. This is one of the things I will consider after merge. IIRC there should only be two fundamental kinds of events: non-destructive adjustment after most edits (which just requires adjustment values), and destructive adjustment after delete (which requires saving a copy)